File: app/src/main/java/io/github/freewebmovement/igniter/activities/exempt/presenter/ExemptAppPresenter.java
package io.github.freewebmovement.igniter.activities.exempt.presenter;

import android.text.TextUtils;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import io.github.freewebmovement.igniter.common.os.Task;
import io.github.freewebmovement.igniter.common.os.Threads;
import io.github.freewebmovement.igniter.activities.exempt.contract.ExemptAppContract;
import io.github.freewebmovement.igniter.persistence.data.AppInfo;
import io.github.freewebmovement.igniter.persistence.data.ExemptAppDataSource;

public class ExemptAppPresenter implements ExemptAppContract.Presenter {
    private final ExemptAppContract.View mView;
    private final ExemptAppDataSource mDataSource;
    private boolean mDirty;
    private boolean mConfigurationChanged;
    private List<AppInfo> mAllAppInfoList;
    private Set<String> mExemptAppPackageNameSet;

    public ExemptAppPresenter(ExemptAppContract.View view, ExemptAppDataSource dataSource) {
        super();
        mView = view;
        mDataSource = dataSource;
        view.setPresenter(this);
    }

    @Override
    public void updateAppInfo(AppInfo appInfo, int position, boolean enable) {
        mDirty = true;
        String packageName = appInfo.getPackageName();
        if (mExemptAppPackageNameSet.contains(packageName)) {
            if (enable) {
                mExemptAppPackageNameSet.remove(packageName);
            }
        } else if (!enable) {
            mExemptAppPackageNameSet.add(packageName);
        }
        appInfo.setEnabled(enable);
    }

    @Override
    public void filterAppsByName(final String name) {
        if (TextUtils.isEmpty(name)) {
            mView.showAppList(mAllAppInfoList);
            return;
        }
        Threads.instance().runOnWorkThread(new Task() {
            @Override
            public void onRun() {
                final List<AppInfo> tmpInfoList = new ArrayList<>();
                for (AppInfo appInfo : mAllAppInfoList) {
                    if (appInfo.getAppName().contains(name)) {
                        tmpInfoList.add(appInfo);
                    }
                }
                Threads.instance().runOnUiThread(() -> mView.showAppList(tmpInfoList));
            }
        });
    }

    @Override
    public void saveExemptAppInfoList() {
        if (!mDirty) {
            mView.showSaveSuccess();
            return;
        }
        mConfigurationChanged = true;
        mView.showLoading();
        Threads.instance().runOnWorkThread(new Task() {
            @Override
            public void onRun() {
                mDataSource.saveExemptAppInfoSet(mExemptAppPackageNameSet);
                mDirty = false;
                Threads.instance().runOnUiThread(() -> {
                    mView.dismissLoading();
                    mView.showSaveSuccess();
                });
            }
        });
    }

    @Override
    public boolean handleBackPressed() {
        if (mDirty) {
            mView.showExitConfirm();
        }
        return mDirty;
    }

    @Override
    public void exit() {
        mView.exit(mConfigurationChanged);
    }

    @Override
    public void start() {
        mView.showLoading();
        Threads.instance().runOnWorkThread(new Task() {
            @Override
            public void onRun() {
                showData();
            }
        });
    }

    private void showData() {
        final List<AppInfo> allAppInfoList = mDataSource.getAllAppInfoList();
        mExemptAppPackageNameSet = mDataSource.loadExemptAppPackageNameSet();
        List<AppInfo> exemptApps = new ArrayList<AppInfo>();
        List<AppInfo> enabledApps = new ArrayList<AppInfo>();
        for (AppInfo appInfo : allAppInfoList) {
            if (mExemptAppPackageNameSet.contains(appInfo.getPackageName())) {
                exemptApps.add(appInfo);
            } else {
                enabledApps.add(appInfo);
                appInfo.setEnabled(true);
            }
        }
        // cluster exempted apps.
        Collections.sort(exemptApps, (o1, o2) -> {
            return o1.getAppName().compareTo(o2.getAppName());
        });

        Collections.sort(enabledApps, (o1, o2) -> {
            return o1.getAppName().compareTo(o2.getAppName());
        });

        enabledApps.addAll(exemptApps);
        mAllAppInfoList = enabledApps;
                Threads.instance().runOnUiThread(() -> {
            mView.showAppList(mAllAppInfoList);
            mView.dismissLoading();
        });
    }
}

File: app/src/main/java/io/github/freewebmovement/igniter/activities/exempt/presenter/ExemptAppPresenter.kt
package io.github.freewebmovement.igniter.activities.exempt.presenter

import android.annotation.SuppressLint
import android.text.TextUtils
import io.github.freewebmovement.igniter.activities.exempt.contract.AppTab
import io.github.freewebmovement.igniter.activities.exempt.contract.ExemptAppContract
import io.github.freewebmovement.igniter.common.os.Task
import io.github.freewebmovement.igniter.common.os.Threads
import io.github.freewebmovement.igniter.constants.DefaultInternationalApps
import io.github.freewebmovement.igniter.persistence.data.AppInfo
import io.github.freewebmovement.igniter.persistence.data.ExemptAppDataSource

class ExemptAppPresenter(
    private val mView: ExemptAppContract.View,
    private val mDataSource: ExemptAppDataSource
) : ExemptAppContract.Presenter {
    private var mDirty = false
    private var mConfigurationChanged = false
    private var mAllAppInfoList: List<AppInfo> = emptyList()
    private var mProxyAppPackageNameSet: MutableSet<String> = mutableSetOf()
    private var mTab = AppTab.NORMAL
    private var mFilterName = ""

    init {
        mView.setPresenter(this)
    }

    override fun updateAppInfo(appInfo: AppInfo, position: Int, enable: Boolean) {
        mDirty = true
        val packageName = appInfo.packageName ?: return
        if (enable) {
            mProxyAppPackageNameSet.add(packageName)
        } else {
            mProxyAppPackageNameSet.remove(packageName)
        }
        appInfo.enabled = enable
        applyFilter()
    }

    override fun filterAppsByName(name: String) {
        mFilterName = name
        applyFilter()
    }

    override fun switchTab(tab: AppTab) {
        mTab = tab
        applyFilter()
    }

    override fun selectAll() {
        mDirty = true
        for (appInfo in currentFilteredList()) {
            appInfo.packageName ?: continue
            mProxyAppPackageNameSet.add(appInfo.packageName!!)
            appInfo.enabled = true
        }
        applyFilter()
    }

    override fun deselectAll() {
        mDirty = true
        for (appInfo in currentFilteredList()) {
            appInfo.packageName ?: continue
            mProxyAppPackageNameSet.remove(appInfo.packageName!!)
            appInfo.enabled = false
        }
        applyFilter()
    }

    override fun saveExemptAppInfoList() {
        if (!mDirty) {
            mView.showSaveSuccess()
            return
        }
        mConfigurationChanged = true
        mView.showSaving()
        Threads.runOnWorkThread(object : Task() {
            override fun onRun() {
                mDataSource.saveExemptAppInfoSet(mProxyAppPackageNameSet)
                mDirty = false
                Threads.runOnUiThread {
                    mView.dismissLoading()
                    mView.showSaveSuccess()
                    // Restarting a running proxy applies the new exempt-app list
                    // to the VPN interface immediately.
                    mView.restartProxyIfRunning()
                }
            }
        })
    }

    override fun handleBackPressed(): Boolean {
        if (mDirty) {
            mView.showExitConfirm()
        }
        return mDirty
    }

    override fun exit() {
        mView.exit(mConfigurationChanged)
    }

    override fun start() {
        mView.showLoading()
        Threads.runOnWorkThread(object : Task() {
            override fun onRun() {
                showData()
            }
        })
    }

    @SuppressLint("NewApi")
    private fun showData() {
        val allAppInfoList = mDataSource.getAllAppInfoList()
        mProxyAppPackageNameSet = mDataSource.loadExemptAppPackageNameSet().toMutableSet()
        val proxyApps = ArrayList<AppInfo>()
        val directApps = ArrayList<AppInfo>()
        for (appInfo in allAppInfoList) {
            appInfo.enabled = mProxyAppPackageNameSet.contains(appInfo.packageName)
            if (appInfo.enabled) {
                proxyApps.add(appInfo)
            } else {
                directApps.add(appInfo)
            }
        }
        proxyApps.sortBy { it.appName }
        directApps.sortBy { it.appName }

        proxyApps.addAll(directApps)
        mAllAppInfoList = proxyApps
        Threads.runOnUiThread {
            mView.showAppList(currentFilteredList())
            mView.dismissLoading()
        }
    }

    private fun currentFilteredList(): List<AppInfo> {
        return mAllAppInfoList.filter { appInfo ->
            val isInternational = DefaultInternationalApps.packageNames.contains(appInfo.packageName)
            val inTab = when (mTab) {
                AppTab.NORMAL -> !appInfo.isSystemApp && !isInternational
                AppTab.SYSTEM -> appInfo.isSystemApp && !isInternational
                AppTab.INTERNATIONAL -> isInternational
            }
            inTab &&
                (TextUtils.isEmpty(mFilterName) || appInfo.appName?.contains(mFilterName, ignoreCase = true) == true)
        }
    }

    private fun applyFilter() {
        Threads.runOnUiThread {
            mView.showAppList(currentFilteredList())
        }
    }
}
